src/jit/stacklevelsetter.cpp

```diff
@@ -0,0 +1,266 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#ifndef LEGACY_BACKEND // This file is ONLY used for the RyuJIT backend that uses the linear scan register allocator
+
+#include "stacklevelsetter.h"
+
+StackLevelSetter::StackLevelSetter(Compiler* compiler)
+    : Phase(compiler, "StackLevelSetter", PHASE_STACK_LEVEL_SETTER)
+    , currentStackLevel(0)
+    , maxStackLevel(0)
+#if !FEATURE_FIXED_OUT_ARGS
+    , framePointerRequired(compiler->codeGen->isFramePointerRequired())
+#endif // !FEATURE_FIXED_OUT_ARGS
+    , memAllocator(compiler, CMK_fgArgInfoPtrArr)
+    , putArgNumSlots(&memAllocator)
+{
+}
+
+//------------------------------------------------------------------------
+// DoPhase: Calculate stack slots numbers for outgoing args.
+//
+// Notes:
+//   For non-x86 platforms it calculates the max number of slots
+//   that calls inside this method can push on the stack.
+//   This value is used for sanity checks in the emitter.
+//
+//   For x86 it also sets throw-helper blocks incoming stack depth and set
+//   set framePointerRequired when it is necessary. These values are used to pop
+//   pushed args when an exception occurs.
+void StackLevelSetter::DoPhase()
+{
+    for (BasicBlock* block = comp->fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+        ProcessBlock(block);
+    }
+#if !FEATURE_FIXED_OUT_ARGS
+
+    if (framePointerRequired && !comp->codeGen->isFramePointerRequired())
+    {
+        JITDUMP("framePointerRequired is not set when it is required\n");
+        comp->codeGen->resetWritePhaseForFramePointerRequired();
+        comp->codeGen->setFramePointerRequired(true);
+    }
+#endif // !FEATURE_FIXED_OUT_ARGS
+    assert(maxStackLevel <= comp->fgGetPtrArgCntMax());
+    if (maxStackLevel != comp->fgGetPtrArgCntMax())
+    {
+        JITDUMP("fgPtrArgCntMax was calculated wrong during the morph, the old value: %u, the right value: %u.\n",
+                comp->fgGetPtrArgCntMax(), maxStackLevel);
+        comp->fgSetPtrArgCntMax(maxStackLevel);
+    }
+}
+
+//------------------------------------------------------------------------
+// ProcessBlock: Do stack level calculations for one block.
+//
+// Notes:
+//   Block starts and ends with an empty outgoing stack.
+//   Nodes in blocks are iterated in the reverse order to memorize put_arg_stk
+//   stack sizes.
+//
+// Arguments:
+//   block - the block to process.
+//
+void StackLevelSetter::ProcessBlock(BasicBlock* block)
+{
+    assert(currentStackLevel == 0);
+    LIR::ReadOnlyRange& range = LIR::AsRange(block);
+    for (auto i = range.rbegin(); i != range.rend(); ++i)
+    {
+        GenTree* node = *i;
+        if (NodePutsOnStack(node))
+        {
+            GenTreePutArgStk* putArg   = static_cast<GenTreePutArgStk*>(node);
+            unsigned          numSlots = putArgNumSlots[putArg];
```

Comment:
Shouldn't you also remove putArg from the hash table here? Otherwise, the hash table only grows, linear with the total number of arguments in the function.

Reply:
Good catch, fixed.